session_cleaner/sublime_session_clean.py

```diff
@@ -0,0 +1,103 @@
+#!/bin/env python3
+"""
+Load the Session.sublime_session file from the Sublime Text 3 data directory
+and remove all of the recent workspaces that refer to projects that no longer
+reside on disk.
+
+If no command line arguments are provided, the script will look for the session
+file in the data directory location standard for the current platform. When a
+command line argument is provided, it's assumed to the location of the Data
+folder to use, which can be relative to the current directory or an absolute
+path.
+
+This creates the new session file (if any) to a temporary file first and then
+renames the session into place. During this process the original session file
+is stored as a backup in a file with the current date and time appended.
+
+This script shouldn't be run while Sublime is actively running, since the
+session information will be written out to disk when Sublime terminates, which
+will cause the changes made by this script to be ignored.
+"""
+
+from datetime import datetime
+import json
+import logging
+import os
+import sys
+
+
+def sublime_data_dir():
+    """
+    Obtain the path to the Sublime Text 3 Data directory. If the script was
+    provided any arguments, the first one is assumed to be the data directory
+    to work with; otherwise the Data directory is selected by platform.
+    """
+    if len(sys.argv) > 1:
+        return sys.argv[0]
+
+    if sys.platform.startswith("linux"):
+        return os.path.expanduser("~/.config/sublime-text-3/")
+    elif sys.platform.startswith("win"):
+        return os.path.expandvars("$APPDATA\\Sublime Text 3\\")
+
+    return os.path.expanduser("~/Library/Application Support/Sublime Text 3/")
+
+
+def clean_session():
+    """
+    Load the session file from the data directory and remove all of the
+    projects that no longer exist on disk, writing the session back.
+
+    This attempts to keep a backup of the existing session file and creates a
+    temporary session first in case things go pear shaped.
+    """
+    data_dir = sublime_data_dir()
+    session_file = os.path.join(data_dir, "Local", "Session.sublime_session")
+    tmp_file = session_file + ".tmp"
+    bkp_file = session_file + datetime.now().strftime(".%Y%m%d_%H%M%S")
+
+    logging.info("Using Data Directory: %s", data_dir)
+    try:
+        with open(session_file, encoding="utf-8") as file:
+            session = json.load(file)
+
+        workspaces = session["workspaces"]["recent_workspaces"]
+        present, missing = [], []
+        for f in workspaces:
+            present.append(f) if os.path.isfile(f) else missing.append(f)
+
+        if len(present) == len(workspaces):
+            return logging.info("No missing projects found")
```

Comment:
if there are no missing projects, would it make sense to bail out here rather than create a backup of the session file and write the unchanged session back to disk?

Reply:
oh sorry, just realized it says `return logging.info`... so it already does this :+1:

Reply:
The return on the logging call should make it leave without doing anything in that case, so it's alread doing that. Would it be more clear to have an explicit return on its own line? I tend to do this a lot but I don't actually know how idiomatic it's considered here.

Reply:
It may be that it's clear enough already, I just somehow missed it while skimming through... Wasn't expecting a log call to be returned I guess. A comment could be enough to clarify it, but yeah, I would definitely not miss a `return` on it's own line :D